lib/Nessy/Client.pm

```diff
@@ -26,47 +26,76 @@ my $MESSAGE_SERIAL = 1;
 sub new {
     my($class, %params) = @_;
 
-    my $api_version = $params{api_version} || $class->_default_api_version;
+    $class->_verify_constructor_params(\%params);
 
-    my $url = $params{url} || Carp::croak('url is a required param');
-
-    my($socket1, $socket2) = IO::Socket->socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC);
-
-    $_->autoflush(1) foreach ($socket1, $socket2);
-
-    my $pid = _fork();
+    my $pid = $class->_fork();
     if ($pid) {
         my $self = bless {}, $class;
-        $self->api_version($api_version);
         $self->pid($pid);
-        $self->default_ttl( $params{default_ttl} || $class->_default_ttl );
-        $self->default_timeout( $params{default_timeout} || $class->_default_timeout );
-        $self->serial_responder_registry({});
-
-        my $watcher = $self->_create_socket_watcher($socket1);
-        $self->socket_watcher($watcher);
-
-        $socket2->close();
-
+        $self->_parent_process_setup(%params);
         return $self;
 
     } elsif (defined $pid) {
-        eval {
-            $socket1->close();
-            my $daemon_class = $class->_daemon_class_name;
-            my $daemon = $daemon_class->new(
-                                url => $url,
-                                client_socket => $socket2,
-                                api_version => $api_version);
-            $daemon->run();
-        };
-        Carp::croak($@) if $@;
+        $class->_run_child_process(%params);
         exit;
     } else {
         Carp::croak("Can't fork: $!");
     }
 }
 
+sub _verify_constructor_params {
+    my($class, $params) = @_;
+
+    $params->{api_version} ||= $class->_default_api_version;
+    $params->{url} || Carp::croak('url is a required param');
+    $params->{'socketpair'} ||= [ $class->_make_socket_pair_for_daemon_comms() ];
```

Comment:
Instead of an arrayref this could be a hashref (or make 0 and 1 names constants) so the usage would be like:

```
$params{socketpair}->{parent}
$params{socketpair}->{child}
```

That may or may not obviate the need for `_close_unused_socket`.